tests_integration/setup_repositories.py

```diff
@@ -0,0 +1,222 @@
+    """
+    The GitHub account for which the setup is done.
+    """
+
+    target_names: set[str]
+    """
+    The names of the repositories that should exist after the preparation (excluding the suffix).
+    """
+
+    suffix: str | None = None
+    """
+    If given, only repositories with the given suffix are considered for deletion and the target names
+    are extended with the suffix.
+    """
+
+    delete_only: bool = False
+    """
+    If True, only delete unnecessary repositories and do not create any new ones.
+    """
+
+
+class RepositoryOwner(Protocol):
+    def create_repo(self, name: str) -> Repository:
+        pass
+
+    def get_repo(self, name: str) -> Repository:
+        pass
+
+    def get_repos(self) -> Iterable[Repository]:
+        pass
+
+
+def get_test_feedstock_names() -> set[str]:
+    """
+    Returns the list of feedstock names that are needed for the integration tests.
+    The names do not include the "-feedstock" suffix.
+    """
+    # note: the trailing "/" is needed to only get the directories,
+    # because of a Python bug this only works in Python 3.11 or later (which is fine)
+    # https://bugs.python.org/issue22276
+    return {path.name for path in DEFINITIONS_DIR.glob("*/")} - IGNORE_FEEDSTOCK_NAMES
+
+
+def _or_empty_set(value: set[str]) -> set[str] | str:
+    """
+    Logging helper function that returns "{}" if the given set is empty.
+    """
+    return value or "{}"
+
+
+def prepare_repositories(
+    owner: RepositoryOwner,
+    owner_name: str,
+    existing_repos: Iterable[Repository],
+    target_names: Iterable[str],
+    delete_only: bool,
+    suffix: str | None = None,
+):
+    """
+    Prepares the repositories of a certain owner for the integration tests.
+    Unnecessary repositories are deleted and missing repositories are created.
+
+    :param owner: The owner of the repositories.
+    :param owner_name: The name of the owner (for logging).
+    :param existing_repos: The existing repositories of the owner.
+    :param target_names: The names of the repositories that should exist after the preparation (excluding the suffix).
+    :param suffix: If given, only repositories with the given suffix are considered for deletion and the target names
+                   are extended with the suffix.
+    :param delete_only: If True, only delete unnecessary repositories and do not create any new ones.
+    """
+    existing_names = {repo.name for repo in existing_repos}
+    target_names = set(target_names)
+
+    if suffix:
+        existing_names = {name for name in existing_names if name.endswith(suffix)}
+        target_names = {name + suffix for name in target_names}
+
+    to_delete = existing_names - target_names
+    to_create = target_names - existing_names
+
+    LOGGER.info(
+        "Deleting the following repositories for %s: %s",
+        owner_name,
+        _or_empty_set(to_delete),
+    )
+    for name in to_delete:
+        owner.get_repo(name).delete()
```

Comment:
My comment for validation and protection against accidental deletion of resources refers to lines like these. I don't care where we add the protection, but it needs to happen before this line is run.

Reply:
see above